tests/safe_mode/check_safe_mode.go

```diff
@@ -0,0 +1,176 @@
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return body, nil
+}
+
+func traceSourceExtract(ev *pb.SyncerBinlogEvent) string {
+	return strings.Split(ev.Base.TraceID, ".")[0]
+}
+
+func binlogPosition(ev *pb.SyncerBinlogEvent) mysql.Position {
+	return mysql.Position{
+		Name: ev.State.CurrentPos.Name,
+		Pos:  ev.State.CurrentPos.Pos,
+	}
+}
+
+func checkSafeMode(obtained, expected bool, ev, startEv, endEv *pb.SyncerBinlogEvent) {
+	if obtained != expected {
+		panic(fmt.Sprintf("safe_mode should be %v, got %v\nev: %v\nstart_ev: %v\nend_ev: %v\n",
+			expected, obtained, ev, startEv, endEv))
+	}
+}
+
+type BinlogEventSlice []*pb.SyncerBinlogEvent
+
+func (c BinlogEventSlice) Len() int {
+	return len(c)
+}
+
+func (c BinlogEventSlice) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
+func (c BinlogEventSlice) Less(i, j int) bool {
+	posi := binlogPosition(c[i])
+	posj := binlogPosition(c[j])
+	return posi.Compare(posj) < 0
+}
+
+func main() {
+	var (
+		raw    []byte
+		err    error
+		offset int
+		limit  int = 10
+		count  int
+	)
+	updateEvents := make(map[string][]*pb.SyncerBinlogEvent)
+	ddlEvents := make(map[string][]*pb.SyncerBinlogEvent)
+	latestDDL := make(map[string][]*pb.SyncerBinlogEvent)
+	for {
+		raw, err = scan_events(offset, limit)
+		if err != nil {
+			panic(errors.ErrorStack(err))
+		}
+		var events [][]tracer.TraceEvent
+		err = json.Unmarshal(raw, &events)
+		if err != nil {
+			panic(err)
+		}
+
+		for _, event_group := range events {
+			for _, e := range event_group {
+				if e.Type == pb.TraceType_BinlogEvent {
+					ev := &pb.SyncerBinlogEvent{}
+					jsonE, _ := json.Marshal(e.Event)
+					err = jsonpb.UnmarshalString(string(jsonE), ev)
+					if err != nil {
+						panic(err)
+					}
+					worker := traceSourceExtract(ev)
+					if ev.OpType == opTypeUpdate {
+						if _, ok2 := updateEvents[worker]; !ok2 {
+							updateEvents[worker] = make([]*pb.SyncerBinlogEvent, 0)
+						}
+						updateEvents[worker] = append(updateEvents[worker], ev)
+					} else if ev.OpType == opTypeDDL {
+						if _, ok2 := ddlEvents[worker]; !ok2 {
+							ddlEvents[worker] = make([]*pb.SyncerBinlogEvent, 0)
+						}
+						ddlEvents[worker] = append(ddlEvents[worker], ev)
+					}
+				}
+			}
+		}
+
+		offset += len(events)
+		if len(events) < limit {
+			break
+		}
+	}
+
+	// sort and get last sharding DDL tracing events
+	for worker, events := range ddlEvents {
+		sort.Sort(BinlogEventSlice(events))
+		latestDDL[worker] = make([]*pb.SyncerBinlogEvent, 2)
+		latestDDL[worker][1] = events[len(events)-1]
+		endPos := binlogPosition(latestDDL[worker][1])
+		for i := len(events) - 2; i >= 0; i-- {
+			pos := binlogPosition(events[i])
+			if pos.Compare(endPos) != 0 {
+				latestDDL[worker][0] = events[i]
+				break
+			}
+		}
+	}
+
+	// check safe mode of each `update` binlog event is reasonable
+	for worker, evs := range latestDDL {
+		startPos := binlogPosition(evs[0])
```

Comment:
is `evs[0]` possible invalid if L153 not executed?

Reply:
added ddl events validation